File: frontend/src/utils/dataGenerator.js
// Normal distribution helper (Box-Muller transform)
const normalRandom = () => {
  const u = 1 - Math.random();
  const v = Math.random();
  return Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
};

export const generateStoreHeatmapData = (width = 50, height = 30, numPoints = 800) => {
  const data = [];
  
  // Define realistic store zones matching the layout
  const zones = [
    // Entrance - High traffic
    { x: 9, y: 7, spread: 3, intensity: 0.95, weight: 0.20, name: 'Entrance' },
    
    // Electronics - Medium-high traffic, longer dwell time
    { x: 41, y: 10, spread: 4, intensity: 0.85, weight: 0.15, name: 'Electronics' },
    
    // Clothing - High traffic, browsing behavior
    { x: 17, y: 16, spread: 5, intensity: 0.80, weight: 0.18, name: 'Clothing' },
    
    // Groceries - Highest traffic, quick movement
    { x: 37, y: 22, spread: 6, intensity: 0.90, weight: 0.22, name: 'Groceries' },
    
    // Checkout - Very high traffic, queue formation
    { x: 13, y: 25, spread: 3, intensity: 0.93, weight: 0.20, name: 'Checkout' },
    
    // Furniture/Home - Lower traffic, longer dwell
    { x: 7, y: 16, spread: 2, intensity: 0.65, weight: 0.05, name: 'Furniture' }
  ];
  
  // Calculate points per zone based on weights
  zones.forEach(zone => {
    const pointsForZone = Math.floor(numPoints * zone.weight);
    
    for (let i = 0; i < pointsForZone; i++) {
      // Use normal distribution for more realistic clustering
      const angle = Math.random() * 2 * Math.PI;
      const radius = Math.abs(normalRandom()) * zone.spread;
      
      const x = Math.max(0, Math.min(width, zone.x + Math.cos(angle) * radius));
      const y = Math.max(0, Math.min(height, zone.y + Math.sin(angle) * radius));
      
      // Add some randomness to intensity
      const intensityVariation = 0.8 + Math.random() * 0.4;
      
      data.push({
        x: x,
        y: y,
        intensity: zone.intensity * intensityVariation,
        zone: zone.name,
        timestamp: new Date().toISOString()
      });
    }
  });
  
  // Add some random movement paths (people walking through aisles)
  const pathPoints = Math.floor(numPoints * 0.05);
  for (let i = 0; i < pathPoints; i++) {
    // Horizontal paths
    if (Math.random() > 0.5) {
      data.push({
        x: Math.random() * width,
        y: 10 + Math.floor(Math.random() * 2) * 10,
        intensity: 0.3 + Math.random() * 0.3,
        zone: 'Aisle',
        timestamp: new Date().toISOString()
      });
    } else {
      // Vertical paths
      data.push({
        x: 15 + Math.floor(Math.random() * 2) * 15,
        y: Math.random() * height,
        intensity: 0.3 + Math.random() * 0.3,
        zone: 'Aisle',
        timestamp: new Date().toISOString()
      });
    }
  }
  
  return data;
};

export const generateCityHeatmapData = () => {
  const cities = [
    { name: 'Downtown Central', lat: 28.6139, lon: 77.2090, demand: 850, population: 125000 },
    { name: 'North District Mall', lat: 28.7041, lon: 77.1025, demand: 620, population: 89000 },
    { name: 'South Hub Plaza', lat: 28.5355, lon: 77.3910, demand: 780, population: 105000 },
    { name: 'East Market Square', lat: 28.6328, lon: 77.2773, demand: 540, population: 72000 },
    { name: 'West Plaza Center', lat: 28.6692, lon: 77.1178, demand: 690, population: 95000 },
    { name: 'Central Square Mall', lat: 28.6289, lon: 77.2065, demand: 920, population: 140000 },
    { name: 'Tech Park Complex', lat: 28.5672, lon: 77.3250, demand: 450, population: 58000 },
    { name: 'Old Town Bazaar', lat: 28.6562, lon: 77.2410, demand: 380, population: 45000 },
    { name: 'Metro Station Hub', lat: 28.6430, lon: 77.2194, demand: 810, population: 118000 },
    { name: 'Airport Road Mall', lat: 28.5562, lon: 77.0999, demand: 720, population: 102000 }
  ];
  
  return cities.map(city => ({
    ...city,
    score: (city.demand * 0.6 + city.population / 1000 * 0.4).toFixed(2),
    footfall: Math.floor(city.demand * 100 + Math.random() * 5000),
    competition: Math.floor(Math.random() * 5) + 1,
    accessibility: (Math.random() * 3 + 7).toFixed(1)
  }));
};

File: backend/data_generator.py
import numpy as np
import random
from typing import List, Dict
from datetime import datetime, timedelta

class DataGenerator:
    @staticmethod
    def generate_store_heatmap(width: int = 50, height: int = 30, num_points: int = 500) -> List[Dict]:
        """Generate simulated store traffic data"""
        hotspots = [
            {"x": 10, "y": 8, "intensity": 0.9, "name": "Entrance"},
            {"x": 40, "y": 12, "intensity": 0.8, "name": "Electronics"},
            {"x": 25, "y": 20, "intensity": 0.7, "name": "Groceries"},
            {"x": 15, "y": 25, "intensity": 0.6, "name": "Checkout"}
        ]
        
        data = []
        for i in range(num_points):
            hotspot = random.choice(hotspots)
            spread = random.random() * 5
            angle = random.random() * 2 * np.pi
            
            x = max(0, min(width, hotspot["x"] + np.cos(angle) * spread))
            y = max(0, min(height, hotspot["y"] + np.sin(angle) * spread))
            
            data.append({
                "id": i,
                "x": float(x),
                "y": float(y),
                "intensity": float(hotspot["intensity"] * (0.5 + random.random() * 0.5)),
                "timestamp": (datetime.now() - timedelta(seconds=random.randint(0, 3600))).isoformat(),
                "zone": hotspot["name"]
            })
        
        return data
    
    @staticmethod
    def generate_city_data() -> List[Dict]:
        """Generate simulated city placement data"""
        cities = [
            {"name": "Downtown", "lat": 28.6139, "lon": 77.2090, "demand": 850, "population": 125000},
            {"name": "North District", "lat": 28.7041, "lon": 77.1025, "demand": 620, "population": 89000},
            {"name": "South Hub", "lat": 28.5355, "lon": 77.3910, "demand": 780, "population": 105000},
            {"name": "East Market", "lat": 28.6328, "lon": 77.2773, "demand": 540, "population": 72000},
            {"name": "West Plaza", "lat": 28.6692, "lon": 77.1178, "demand": 690, "population": 95000},
            {"name": "Central Square", "lat": 28.6289, "lon": 77.2065, "demand": 920, "population": 140000},
            {"name": "Tech Park", "lat": 28.5672, "lon": 77.3250, "demand": 450, "population": 58000},
            {"name": "Old Town", "lat": 28.6562, "lon": 77.2410, "demand": 380, "population": 45000}
        ]
        
        for city in cities:
            # Calculate score: 60% demand + 40% population density
            city["score"] = round(city["demand"] * 0.6 + city["population"] / 1000 * 0.4, 2)
            city["traffic_density"] = round(city["demand"] / (city["population"] / 1000), 2)
        
        return sorted(cities, key=lambda x: x["score"], reverse=True)
    
    @staticmethod
    def generate_analytics() -> Dict:
        """Generate analytics summary"""
        return {
            "total_visitors": random.randint(1000, 1500),
            "avg_dwell_time": round(random.uniform(15, 25), 1),
            "conversion_rate": round(random.uniform(30, 40), 1),
            "peak_hours": [
                {"hour": "10-11 AM", "visitors": random.randint(120, 160)},
                {"hour": "12-1 PM", "visitors": random.randint(180, 220)},
                {"hour": "2-3 PM", "visitors": random.randint(150, 180)},
                {"hour": "5-6 PM", "visitors": random.randint(200, 240)},
                {"hour": "7-8 PM", "visitors": random.randint(170, 200)}
            ]
        }